File: src/main/java/earth/terrarium/tempad/client/components/InformationPanel.java
package earth.terrarium.tempad.client.components;

import com.teamresourceful.resourcefullib.client.components.selection.SelectionList;
import earth.terrarium.tempad.client.config.TempadClientConfig;
import earth.terrarium.tempad.common.Tempad;
import earth.terrarium.tempad.api.locations.LocationData;
import net.minecraft.core.GlobalPos;
import net.minecraft.network.chat.Component;
import net.minecraft.util.Mth;

import java.util.List;

public class InformationPanel extends SelectionList<TextEntry> {

    private static final TextEntry NO_SELECTION_LINE_1 = new TextEntry(Component.translatable("gui." + Tempad.MODID + ".no_selection.first_line"));
    private static final TextEntry NO_SELECTION_LINE_2 = new TextEntry(Component.translatable("gui." + Tempad.MODID + ".no_selection.second_line"));
    private static final List<TextEntry> NO_SELECTION = List.of(NO_SELECTION_LINE_1, NO_SELECTION_LINE_2);

    private static final String LOCATION_X = "gui." + Tempad.MODID + ".x";
    private static final String LOCATION_Y = "gui." + Tempad.MODID + ".y";
    private static final String LOCATION_Z = "gui." + Tempad.MODID + ".z";
    private static final String ANGLE = "gui." + Tempad.MODID + ".angle";
    private static final String LOCATION_DIMENSION = "gui." + Tempad.MODID + ".dimension";
    private final int color;
    private final boolean shadow;

    public InformationPanel(int x, int y, int width, int height) {
        this(x, y, width, height, TempadClientConfig.color, true);
    }

    public InformationPanel(int x, int y, int width, int height, int color, boolean shadow) {
        super(x, y, width, height, 10, entry -> {});
        this.color = color;
        this.shadow = shadow;
    }

    public void updateBlank() {
        updateBlank(true);
    }

    public void updateBlank(boolean showNoSelection) {
        if (showNoSelection) {
            updateEntries(NO_SELECTION);
        } else {
            updateEntries(List.of());
        }
    }

    public void update(LocationData location) {
        updateEntries(List.of(
            new TextEntry(Component.literal(location.name()), color, shadow),
            new TextEntry(Component.translatable(LOCATION_X, Mth.floor(location.blockPos().getX())), color, shadow),
            new TextEntry(Component.translatable(LOCATION_Y, Mth.floor(location.blockPos().getY())), color, shadow),
            new TextEntry(Component.translatable(LOCATION_Z, Mth.floor(location.blockPos().getZ())), color, shadow),
            new TextEntry(Component.translatable(ANGLE, location.angle()), color, shadow),
            new TextEntry(Component.translatable(LOCATION_DIMENSION, Component.translatable(location.levelKey().location().toLanguageKey("dimension"))), color, shadow)
            )
        );
    }

    public void updateNameless(GlobalPos pos) {
        updateEntries(List.of(
            new TextEntry(Component.translatable(LOCATION_X, Mth.floor(pos.pos().getX())), color, shadow),
            new TextEntry(Component.translatable(LOCATION_Y, Mth.floor(pos.pos().getY())), color, shadow),
            new TextEntry(Component.translatable(LOCATION_Z, Mth.floor(pos.pos().getZ())), color, shadow),
            new TextEntry(Component.translatable(LOCATION_DIMENSION, Component.translatable(pos.dimension().location().toLanguageKey("dimension"))), color, shadow))
        );
    }
}

File: src/main/kotlin/earth/terrarium/tempad/client/widgets/InformationPanel.kt
package earth.terrarium.tempad.client.widgets

import earth.terrarium.olympus.client.components.base.ListWidget
import earth.terrarium.tempad.Tempad
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.components.MultiLineTextWidget
import net.minecraft.client.gui.components.StringWidget
import net.minecraft.client.gui.navigation.ScreenRectangle
import net.minecraft.network.chat.Component

class InformationPanel(font: Font, width: Int, height: Int) : ListWidget(width, height) {
    private val NO_SELECTION: MultilineEntry = MultilineEntry(Component.translatable(("app.${Tempad.MOD_ID}.teleport.no_selection")), font).apply {
        setWidth(width)
        setMaxWidth(width - 8)
    }

    init {
        gap = 2
        set(listOf(NO_SELECTION))
    }

    fun update(display: List<Component>) {
        set(display.map { StringEntry(it) })
    }

    fun clearLines() {
        set(listOf(NO_SELECTION))
    }
}

class StringEntry(
    text: Component,
) : StringWidget(text, Minecraft.getInstance().font), ListWidget.Item {
    override fun getRectangle(): ScreenRectangle {
        return super<StringWidget>.getRectangle()
    }

    init {
        setColor(Tempad.ORANGE.value)
        alignLeft()
    }

    override fun getX(): Int = super.getX() + 4
    override fun getY(): Int = super.getY() + 4
    override fun getWidth(): Int = super<StringWidget>.width - 4
}

class MultilineEntry(text: Component, font: Font) : MultiLineTextWidget(text, font), ListWidget.Item {
    override fun getRectangle(): ScreenRectangle {
        return super<MultiLineTextWidget>.getRectangle()
    }

    init {
        setColor(Tempad.ORANGE.value)
    }

    override fun getX(): Int = super.getX() + 4
    override fun getY(): Int = super.getY() + 4
}
